plugins/ltac2/tac2compile.ml

```diff
@@ -0,0 +1,1012 @@
+   depends on nothing, the second may depend on the first, etc.
+
+   This function adds [kn] and all its dependencies to the list. *)
+let rec get_dependencies ((visited, skipped_mut, knl) as acc) kn =
+  if KNset.mem kn visited then acc
+  else
+    let data = Tac2env.interp_global kn in
+    let skipped_mut =
+      if data.gdata_mutable
+      then KNset.add kn skipped_mut
+      else skipped_mut
+    in
+    let kndeps = get_expr_deps data.gdata_expr in
+    KNset.fold (fun kn acc -> get_dependencies acc kn)
+      kndeps
+      (KNset.add kn visited, skipped_mut, kn :: knl)
+
+let warn_skipped_mut = CWarnings.create ~name:"tac2compile-skipped-mutable" ~category:"ltac2"
+    (fun skipped_mut ->
+       str "Skipped compilation of mutable definitions" ++ spc() ++
+       prlist_with_sep spc Tac2print.pr_tacref (KNset.elements skipped_mut))
+
+let get_recursive_kns knl =
+  let _, skipped_mut, knl = List.fold_left get_dependencies (KNset.empty, KNset.empty, []) knl in
+  let () = if not (KNset.is_empty skipped_mut) then warn_skipped_mut skipped_mut in
+  knl
+
+let my_temp_dir = ref None
+
+let force_temp_dir () = match !my_temp_dir with
+  | Some d -> d
+  | None ->
+    let d = CUnix.mktemp_dir "tac2compile_" "" in
+    my_temp_dir := Some d;
+    d
+
+let () = at_exit (fun () ->
+    if not (CDebug.get_flag debug_flag) && Option.has_some !my_temp_dir then
+      try
+        let d = Option.get !my_temp_dir in
+        Array.iter (fun f -> Sys.remove (Filename.concat d f)) (Sys.readdir d);
+        Unix.rmdir d
+      with e ->
+        Feedback.msg_warning
+          Pp.(str "tac2compile: failed to cleanup: " ++
+              str(Printexc.to_string e) ++ fnl()))
+
+let get_ml_filename () =
+  let temp_dir = force_temp_dir () in
+  let filename, ch = Filename.open_temp_file ~temp_dir "f" ".ml" in
+  let prefix = Filename.chop_extension (Filename.basename filename) ^ "." in
+  filename, ch, prefix
+
+let error_compiler_failed e =
+  let msg = match e with
+  | Inl (Unix.WEXITED 127) -> Pp.(strbrk "The OCaml compiler was not found. Make sure it is installed, together with findlib.")
+  | Inl (Unix.WEXITED n) ->
+     Pp.(strbrk "Ltac2 compiler exited with status" ++ str" " ++ int n
+         ++ strbrk (if n = 2 then " (in case of stack overflow, increasing stack size (typically with \"ulimit -s\") often helps)" else ""))
+  | Inl (Unix.WSIGNALED n) -> Pp.(strbrk "Ltac2 compiler killed by signal" ++ str" " ++ int n)
+  | Inl (Unix.WSTOPPED n) -> Pp.(strbrk "Ltac2 compiler stopped by signal" ++ str" " ++ int n)
+  | Inr e -> Pp.(strbrk "Ltac2 compiler failed with error: " ++ strbrk (Unix.error_message e))
+  in
+  CErrors.user_err msg
+
+let include_dirs () =
+  (* TODO make this work in -boot / dev shim mode *)
+  let open Boot.Env in
+  let env = init () in
+  (* engine for Proofview, kernel for Names *)
+  List.map (fun x -> Path.to_string (native_cmi env x))
+    [ "kernel"; "engine"; "plugins/ltac2" ]
```

Comment:
This works with an installed Coq on my machine.
As noted in the comment it doesn't work if you want to use in the stdlib (-boot) or with the dev shims.

Reply:
at least it's not some hardcoded paths ;)

Reply:
We can add the dep to the dev shims. For the -boot thing, @Alizter and I were discussing recently how to continue #17317 as to improve this, we could start by having `Boot.Env.init` return an option type.

This way, if the above fail, you can workaround it (for example by assuming a configure'd env)